rpc/metadata.go

```diff
@@ -0,0 +1,34 @@
+package rpc
+
+import (
+	"context"
+	"net/http"
+)
+
+type mdHeaderKey struct{}
+
+// NewContextWithHeaders is used to attach http headers into the context
+func NewContextWithHeaders(ctx context.Context, header http.Header) context.Context {
+	return context.WithValue(ctx, mdHeaderKey{}, header)
+}
+
+// HeadersFromContext is used to extract http headers from the context
+func HeadersFromContext(ctx context.Context) http.Header {
```

Comment:
In go-ethereum, we have to be mindful about backwards-compatibility. Every time we add a method which is public, it increases the maintainence-surface: because once we've added a public method, it is part of the public API, and very difficult to remove.

So we never "proactively" add a `HeadersFromContext` when we only strictly need a `headersFromContext`. We only ever make it public if we absolutely have to. And in this case, we do not absolutely have to, as far as I can tell. The caller can set headers in the constructor.  

```suggestion
// headersFromContext is used to extract http headers from the context
func headersFromContext(ctx context.Context) http.Header {
```

Reply:
IMO it's better to just inline this method into `addHeadersFromContext`

Reply:
@holiman  I agree with your consideration of backward compatibility.
However, the user's call link may be very complex. During the transmission of context layer by layer, `NewContextWithHeaders` can only play the role of `overwriting`, but not `append`. Users at any layer need to check the code to see whether the upper level layer has set the `mdHeaderKey`, otherwise it will be overwritten, resulting in unpredictable consequences.
The design here refers: https://github.com/grpc/grpc-go/blob/master/metadata/metadata.go#L179-L196.
When reading your suggestion, I considered deleting `HeadersFromContext` and replacing `NewContextWithHeaders` with `AppendHeadersIntoContext`. It can solve the problem I mentioned above and reduce the mental burden of users, but its problem is that users will not be able to empty/delete the http.Headers, Although the requirement to forcibly empty/delete headers at a certain layer of code does not seem so common (except for debugging).
Of course, setting `HeadersFromContext` to private is acceptable to me, because `NewContextWithHeaders` is enough for my current needs. 
Do you have any further ideas? For example, replace `NewContextWithHeaders` with `AppendHeadersIntoContext` ?